examples/Tutorial 4 - 3SAT Solver.ipynb

```diff
@@ -0,0 +1,268 @@
+    "        for i in range(len(self.clauses)):\n",
+    "            graph.add_edges_from([(3*i,3*i+1), (3*i+1, 3*i+2), (3*i, 3*i+2)])\n",
+    "            # add edges between each node and its complementaries\n",
+    "            for j in range(3):\n",
+    "                for node in nodes[self.clauses[i].terms[j].variable + (\"F\" if self.clauses[i].terms[j].positive else \"T\")]:\n",
+    "                    graph.add_edge(3*i + j, node)\n",
+    "            \n",
+    "            # add the new nodes to the cache\n",
+    "            for j in range(3):\n",
+    "                nodes[self.clauses[i].terms[j].variable + (\"T\" if self.clauses[i].terms[j].positive else \"F\")].append(3*i+j)\n",
+    "                \n",
+    "        return MISInstance(graph)\n",
+    "            \n",
+    "    \n",
+    "\n",
+    "    def rebuild_result_from_mis(self, solutions: list[MISSolution]) -> dict[str, bool] | None:\n",
+    "        \"\"\"\n",
+    "        Search in a list of MISSolution if any of them is a possible solution to 3SAT.\n",
+    "        \"\"\"\n",
+    "        if solutions is None:\n",
+    "            return None\n",
+    "        \n",
+    "        for solution in solutions:\n",
+    "            if len(solution.nodes) != len(self.clauses): \n",
+    "                continue # skip if the number of nodes isn't enough to be a solution \n",
+    "            # otherwise, we have a solution\n",
+    "            result = dict()\n",
+    "            for variable in self._variables:\n",
+    "                result[variable] = False # set the default of every variable to False in case the current solution makes them free\n",
+    "                \n",
+    "            for node in solution.nodes:\n",
+    "                term = self.clauses[node//3].terms[node%3] # since each node has index equal to 3*clause + term\n",
+    "                result[term.variable] = term.positive\n",
+    "            \n",
+    "            return result\n",
+    "                \n",
+    "        return None\n",
+    "    \n",
+    "    "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 87,
+   "id": "c4dcb84e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Possible solution :  {'X1': True, 'X2': False, 'X3': False}\n",
+      "True\n"
+     ]
+    }
+   ],
+   "source": [
+    "\n",
+    "# Testing our class\n",
+    "\n",
+    "three_sat_instance = ThreeSATInstance(clauses=[\n",
+    "   ThreeSATClause([\n",
+    "      ThreeSATTerm(\"X1\", positive=True),\n",
+    "      ThreeSATTerm(\"X2\", positive=True),\n",
+    "      ThreeSATTerm(\"X3\", positive=True),\n",
+    "   ]),\n",
+    "   ThreeSATClause([\n",
+    "      ThreeSATTerm(\"X1\", positive=True),\n",
+    "      ThreeSATTerm(\"X2\", positive=False),\n",
+    "      ThreeSATTerm(\"X3\", positive=False),\n",
+    "   ]),\n",
+    "])\n",
+    "\n",
+    "config = SolverConfig()\n",
```

Comment:
Before testing this, could you check that a few hand-written `values` have the expected result with `eval`?